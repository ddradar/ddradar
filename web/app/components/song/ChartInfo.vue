<script setup lang="ts">
import type { StepChart } from '@ddradar/core'
import { difficultyMap, playStyleMap } from '@ddradar/core'

import type { AccordionItem } from '#ui/types'

const props = defineProps<{ chart: StepChart }>()

const { t } = useI18n()
const title = computed(() => {
  const playStyleText = playStyleMap.get(props.chart.playStyle)!
  const difficultyName = difficultyMap.get(props.chart.difficulty)!
  const shortPlayStyle = `${playStyleText[0]}P` as 'SP' | 'DP'
  return `${shortPlayStyle}-${difficultyName} (${props.chart.level})` as const
})
const color = computed(() => getChartColor(props.chart.difficulty))
const items = computed<AccordionItem[]>(() => [
  { label: title.value, slot: 'chart' },
  { label: t('score'), slot: 'score' },
])
</script>

<template>
  <UAccordion :color="color" :items="items">
    <template #chart>
      <div class="grid grid-cols-2 h-300">
        <div class="m-2">
          <ul>
            <li><em>Notes</em>: {{ chart.notes }}</li>
            <li><em>Freeze Arrow</em>: {{ chart.freezeArrow }}</li>
            <li><em>Shock Arrow</em>: {{ chart.shockArrow }}</li>
            <br />
          </ul>
        </div>
      </div>
    </template>
  </UAccordion>
</template>

<i18n lang="json">
{
  "ja": {
    "score": "スコアランキング",
    "chart": "譜面情報"
  },
  "en": {
    "score": "Score Board",
    "chart": "Chart Info"
  }
}
</i18n>
