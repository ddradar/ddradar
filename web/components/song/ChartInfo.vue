<script setup lang="ts">
import type { CourseChartSchema, StepChartSchema } from '@ddradar/core'
import { difficultyMap, playStyleMap } from '@ddradar/core'

import type { AccordionItem } from '#ui/types'

const props = defineProps<{ chart: CourseChartSchema | StepChartSchema }>()

const { t } = useI18n()
const title = computed(() => {
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  const playStyleText = playStyleMap.get(props.chart.playStyle)!
  const difficultyName = difficultyMap.get(props.chart.difficulty)!
  /* eslint-enable @typescript-eslint/no-non-null-assertion */
  const shortPlayStyle = `${playStyleText[0]}P` as 'SP' | 'DP'
  return `${shortPlayStyle}-${difficultyName} (${props.chart.level})` as const
})
const color = computed(() => getChartColor(props.chart.difficulty))
const items = computed<AccordionItem[]>(() => [
  { label: t('score'), slot: 'score' },
  { label: t('chart'), slot: 'chart' },
  ...(isCourseChart(props.chart)
    ? [{ label: t('course'), slot: 'course' }]
    : []),
])
</script>

<template>
  <UCard
    :ui="{ body: { base: '', padding: 'px-0 py-2 sm:px-0 sm:py-2' } }"
    :color="color"
  >
    <template #header>
      {{ title }}
    </template>
    <UAccordion :color="color" :items="items">
      <template #chart>
        <ul>
          <li><em>Notes</em>: {{ chart.notes }}</li>
          <li><em>Freeze Arrow</em>: {{ chart.freezeArrow }}</li>
          <li><em>Shock Arrow</em>: {{ chart.shockArrow }}</li>
        </ul>
      </template>
    </UAccordion>
  </UCard>
</template>

<i18n lang="json">
{
  "ja": {
    "score": "スコアランキング",
    "chart": "譜面情報",
    "course": "コース曲目"
  },
  "en": {
    "score": "Score Board",
    "chart": "Chart Info",
    "course": "Course Order"
  }
}
</i18n>
