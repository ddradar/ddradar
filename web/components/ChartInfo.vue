<template>
  <section
    class="column is-half-tablet is-one-third-desktop is-one-quarter-widescreen"
  >
    <Card title="Chart Info" :variant="cardType" collapsible>
      <div class="card-content">
        <div class="content">
          <ul>
            <li><em>Notes</em>: {{ chart.notes }}</li>
            <li><em>Freeze Arrow</em>: {{ chart.freezeArrow }}</li>
            <li><em>Shock Arrow</em>: {{ chart.shockArrow }}</li>
          </ul>
          <ol v-if="isCourse(chart)">
            <li v-for="o in chart.order" :key="o.songId">
              <nuxt-link
                :to="`/songs/${o.songId}#${o.playStyle}-${o.difficulty}`"
              >
                {{ o.songName }}
              </nuxt-link>
              {{ getChartTitle(o) }}
            </li>
          </ol>
        </div>
        <div v-if="!isCourse(chart)" class="table-container">
          <table class="table">
            <thead>
              <tr>
                <th><abbr title="STREAM">STR</abbr></th>
                <th><abbr title="VOLTAGE">VOL</abbr></th>
                <th>AIR</th>
                <th><abbr title="FREEZE">FRE</abbr></th>
                <th><abbr title="CHAOS">CHA</abbr></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>{{ chart.stream }}</td>
                <td>{{ chart.voltage }}</td>
                <td>{{ chart.air }}</td>
                <td>{{ chart.freeze }}</td>
                <td>{{ chart.chaos }}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </Card>
  </section>
</template>

<script lang="ts" setup>
import { Song } from '@ddradar/core'
import { computed } from 'vue'

import Card from '~/components/Card.vue'
import type { CourseInfo } from '~/server/api/v1/courses/[id].get'
import type { SongInfo } from '~/server/api/v1/songs/[id].get'
import { difficultyMap, getChartTitle } from '~/src/song'

type CourseChart = CourseInfo['charts'][number]
type Chart = SongInfo['charts'][number] | CourseChart

interface ChartInfoProps {
  chart: Chart
}

const props = defineProps<ChartInfoProps>()

const cardType = computed(
  () =>
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    difficultyMap
      .get(props.chart.difficulty)!
      .toLowerCase() as Lowercase<Song.DifficultyName>
)

const isCourse = (chart: Chart): chart is CourseChart =>
  Array.isArray((chart as CourseChart).order)
</script>
