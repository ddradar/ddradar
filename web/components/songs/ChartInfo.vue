<template>
  <section
    class="column is-half-tablet is-one-third-desktop is-one-quarter-widescreen"
  >
    <ScoreBoard :song-id="songId" :is-course="isCourse" :chart="chart" />
    <CollapsibleCard title="Chart Info" :variant="cardType" collapsible>
      <div class="card-content">
        <div class="content">
          <ul>
            <li><em>Notes</em>: {{ chart.notes }}</li>
            <li><em>Freeze Arrow</em>: {{ chart.freezeArrow }}</li>
            <li><em>Shock Arrow</em>: {{ chart.shockArrow }}</li>
          </ul>
          <ol v-if="isCourse">
            <li
              v-for="o in (chart as CourseInfo['charts'][number]).order"
              :key="o.songId"
            >
              <nuxt-link
                :to="`/songs/${o.songId}#${o.playStyle}-${o.difficulty}`"
              >
                {{ o.songName }}
              </nuxt-link>
              {{ getChartTitle(o) }}
            </li>
          </ol>
        </div>
        <div v-if="!isCourse" class="table-container">
          <table class="table">
            <thead>
              <tr>
                <th><abbr title="STREAM">STR</abbr></th>
                <th><abbr title="VOLTAGE">VOL</abbr></th>
                <th>AIR</th>
                <th><abbr title="FREEZE">FRE</abbr></th>
                <th><abbr title="CHAOS">CHA</abbr></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>{{ (chart as SongInfo['charts'][number]).stream }}</td>
                <td>{{ (chart as SongInfo['charts'][number]).voltage }}</td>
                <td>{{ (chart as SongInfo['charts'][number]).air }}</td>
                <td>{{ (chart as SongInfo['charts'][number]).freeze }}</td>
                <td>{{ (chart as SongInfo['charts'][number]).chaos }}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </CollapsibleCard>
  </section>
</template>

<script lang="ts" setup>
import { difficultyMap } from '@ddradar/core'

import CollapsibleCard from '~~/components/CollapsibleCard.vue'
import ScoreBoard from '~~/components/songs/ScoreBoard.vue'
import type { CourseInfo } from '~~/server/api/v1/courses/[id].get'
import type { SongInfo } from '~~/server/api/v1/songs/[id].get'
import { getChartTitle } from '~~/utils/song'

type Chart = SongInfo['charts'][number] | CourseInfo['charts'][number]

interface ChartInfoProps {
  songId: SongInfo['id']
  chart: Chart
}

const props = defineProps<ChartInfoProps>()

const cardType = computed(() =>
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  difficultyMap.get(props.chart.difficulty)!.toLowerCase()
)
const isCourse = computed(() =>
  Array.isArray((props.chart as CourseInfo['charts'][number]).order)
)
</script>
