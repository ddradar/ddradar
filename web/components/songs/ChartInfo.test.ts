import { testCourseData, testSongData } from '@ddradar/core/test/data'
import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { mount, RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'

import ChartInfo from '~~/components/songs/ChartInfo.vue'

describe('components/songs/ChartInfo.vue', () => {
  test.each([
    { songId: testSongData.id, chart: testSongData.charts[0] },
    { songId: testCourseData.id, chart: testCourseData.charts[1] },
  ])('(%o) snapshot test', props => {
    const wrapper = mount(ChartInfo, {
      props,
      global: {
        plugins: [[Oruga, bulmaConfig]],
        stubs: { NuxtLink: RouterLinkStub, ScoreBoard: true },
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
