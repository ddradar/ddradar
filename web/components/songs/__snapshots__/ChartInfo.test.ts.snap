// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/songs/ChartInfo.vue > ({
  songId: '06loOQ0DQb0DqbOibl6qO81qlIdoP9DI',
  chart: {
    playStyle: 1,
    difficulty: 0,
    level: 4,
    notes: 138,
    freezeArrow: 0,
    shockArrow: 0,
    stream: 29,
    voltage: 22,
    air: 5,
    freeze: 0,
    chaos: 0
  }
}) snapshot test 1`] = `
<section
  class="column is-half-tablet is-one-third-desktop is-one-quarter-widescreen"
>
  <score-board-stub
    chart="[object Object]"
    songid="06loOQ0DQb0DqbOibl6qO81qlIdoP9DI"
  />
  <div
    class="o-clps card"
    data-v-a8d65a31=""
  >
    <div
      class="o-clps__trigger"
    >
      <header
        aria-controls="Chart Info"
        class="card-header is-beginner"
        data-v-a8d65a31=""
        role="button"
      >
        <div
          class="card-header-title"
          data-v-a8d65a31=""
        >
          Chart Info
        </div>
        <a
          class="card-header-icon"
          data-v-a8d65a31=""
        >
          <span
            class="icon"
            data-v-a8d65a31=""
          >
            <i
              class="mdi mdi-menu-down mdi-24px"
            />
          </span>
        </a>
      </header>
    </div>
    <transition-stub
      appear="false"
      css="true"
      name="slide"
      persisted="false"
    >
      <div
        class="o-clps__content"
        id="Chart Info"
        style="display: none;"
      >
        
        <div
          class="card-content"
        >
          <div
            class="content"
          >
            <ul>
              <li>
                <em>
                  Notes
                </em>
                : 138
              </li>
              <li>
                <em>
                  Freeze Arrow
                </em>
                : 0
              </li>
              <li>
                <em>
                  Shock Arrow
                </em>
                : 0
              </li>
            </ul>
            <!--v-if-->
          </div>
          <div
            class="table-container"
          >
            <table
              class="table"
            >
              <thead>
                <tr>
                  <th>
                    <abbr
                      title="STREAM"
                    >
                      STR
                    </abbr>
                  </th>
                  <th>
                    <abbr
                      title="VOLTAGE"
                    >
                      VOL
                    </abbr>
                  </th>
                  <th>
                    AIR
                  </th>
                  <th>
                    <abbr
                      title="FREEZE"
                    >
                      FRE
                    </abbr>
                  </th>
                  <th>
                    <abbr
                      title="CHAOS"
                    >
                      CHA
                    </abbr>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    29
                  </td>
                  <td>
                    22
                  </td>
                  <td>
                    5
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        
      </div>
    </transition-stub>
  </div>
</section>
`;

exports[`components/songs/ChartInfo.vue > ({
  songId: 'I90bQ81P1blOPIdd9PPl6I9D8DQ1dIob',
  chart: {
    playStyle: 1,
    difficulty: 1,
    level: 8,
    notes: 782,
    freezeArrow: 36,
    shockArrow: 0,
    order: [
      {
        songId: 'PdPdbdQ1lDI18O90Q0li8QI0bo99bidd',
        songName: 'This Beat Is.....',
        playStyle: 1,
        difficulty: 1,
        level: 4
      },
      {
        songId: 'PPQb966D6Ql6Q8iPo68810PdDiI6D0bl',
        songName: 'Waiting',
        playStyle: 1,
        difficulty: 1,
        level: 6
      },
      {
        songId: '6iO8d9ld00I6lDIO6qbOldP99oqloqb9',
        songName: 'Dead Heat',
        playStyle: 1,
        difficulty: 1,
        level: 6
      },
      {
        songId: '1iDPo19Pb9ooi8OqiidbqPb06DDdqiqo',
        songName: 'Neverland',
        playStyle: 1,
        difficulty: 1,
        level: 8
      },
      [length]: 4
    ]
  }
}) snapshot test 1`] = `
<section
  class="column is-half-tablet is-one-third-desktop is-one-quarter-widescreen"
>
  <score-board-stub
    chart="[object Object]"
    songid="I90bQ81P1blOPIdd9PPl6I9D8DQ1dIob"
  />
  <div
    class="o-clps card"
    data-v-a8d65a31=""
  >
    <div
      class="o-clps__trigger"
    >
      <header
        aria-controls="Chart Info"
        class="card-header is-basic"
        data-v-a8d65a31=""
        role="button"
      >
        <div
          class="card-header-title"
          data-v-a8d65a31=""
        >
          Chart Info
        </div>
        <a
          class="card-header-icon"
          data-v-a8d65a31=""
        >
          <span
            class="icon"
            data-v-a8d65a31=""
          >
            <i
              class="mdi mdi-menu-down mdi-24px"
            />
          </span>
        </a>
      </header>
    </div>
    <transition-stub
      appear="false"
      css="true"
      name="slide"
      persisted="false"
    >
      <div
        class="o-clps__content"
        id="Chart Info"
        style="display: none;"
      >
        
        <div
          class="card-content"
        >
          <div
            class="content"
          >
            <ul>
              <li>
                <em>
                  Notes
                </em>
                : 782
              </li>
              <li>
                <em>
                  Freeze Arrow
                </em>
                : 36
              </li>
              <li>
                <em>
                  Shock Arrow
                </em>
                : 0
              </li>
            </ul>
            <ol>
              
              <li>
                <a>
                  This Beat Is.....
                </a>
                 SP-BASIC (4)
              </li>
              <li>
                <a>
                  Waiting
                </a>
                 SP-BASIC (6)
              </li>
              <li>
                <a>
                  Dead Heat
                </a>
                 SP-BASIC (6)
              </li>
              <li>
                <a>
                  Neverland
                </a>
                 SP-BASIC (8)
              </li>
              
            </ol>
          </div>
          <!--v-if-->
        </div>
        
      </div>
    </transition-stub>
  </div>
</section>
`;
